### ISSUE TITLE
Hang in batch `Command`

### ISSUE THREAD

**Issue by finestructure** [vapor/console-kit#179]
This is a follow-up of a [Discord discussion](https://discord.com/channels/431917998102675485/519613337638797315/1097832913694883840), FYI @0xTim !

----

We're seeing the following error in one of our Vapor batch jobs every once in a while (a few times per month, maybe once a week - hard to quantify precisely):

```
│   [ ERROR ] Connection request timed out. This might indicate a connection deadlock in your application. If you're running long running requests, consider increasing your connection timeout. [component: server, database-id: psql]                │
```

When this message occurs, the job HANGS - i.e. it locks up and does not terminate, preventing all further processing.

In normal operation the job spins up and typically processes for 3-6 seconds, running a number of queries against the Github APIs (we're processing batches of 25 Swift packages, updating Github metadata in the Swift Package Index).

I've done a little research and there seem to be a couple of cases where this can happen:

1) too many db connections
2) an error [handling problem in networking code](https://discord.com/channels/431917998102675485/519613337638797315/1071481521627742280)

I suspect it's 2). Last I checked we're not exhausing our db connections and our db utilization across our two envs is <15% and <10%.

I believe this issue first started happening after moving to async/await. This may be due to how we're launching async processes from a `Command` via our own [`AyncCommand`](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/main/Sources/App/Commands/AsyncCommand.swift). (We previously used a semaphore-based implementation which also encountered hangs.)

Our logs show no unusual activity around the error, although we've not yet had it happen with `DEBUG` logging enabled.

----

We're also tracking this issue here: https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2227, and I'll be adding more observations as they happen there.

**Comment by finestructure** [comment]
Just happened with debug logs, attached.

One thing that I notice is that there's an exactly 10s gap between the last DEBUG log and the `Connection request timed out` error. That looks like a default timeout and perhaps the length can give a hint which part of the system is timing out (network vs db connection - if that's even a debate, I don't know).

[hang-prod-2023-04-27.logs.zip](https://github.com/vapor/console-kit/files/11347639/hang-prod-2023-04-27.logs.zip)

**Comment by finestructure** [comment]
Ok, I actually have a local reproducer for this now. This happens because a synchronous part of our task is running too long, a `git clone` run in a shell. I can also force the issue more reliably by adding a `sleep(11)` (> the 10s timeout) to the synchronous part.

I'm unsure how to best deal with this though. This needs to be run as part of the task as a whole but it's not clear to me how to offload it to unblock the ELF and resume processing on the ELF once it's done.

I guess I could increase the timeout but to some degree that's just kicking the can down the road until the next bigger repo / network glitch comes along.

I'd like to avoid the hangs, as they require manual restarting of the job.

**Comment by finestructure** [comment]
Ah, that's exactly what I was going to try and figure out tomorrow - how would I offload this properly. Thanks a lot for the pointer!

**Comment by finestructure** [comment]
So using `application.threadPool` doesn't change anything. I may be holding it wrong but while constructing a standalone app to make it easier to discuss I found that the issue seems to be related to db transactions.

If my workload is

```
                        try await database.transaction { tx in
                            print("sleeping")
                            sleep(11)
                        }
```

the `connection deadlock` error is triggered. If it's simply

```
                        print("sleeping")
                        sleep(11)
```

it isn't.

Note that there's no actual db activity involved at all. The example runs without a db.

I may have oversimplified the example and be triggering something that only looks like our issue but it certainly looks like what we're seeing in the full app (which is also using a transaction).

The reproducer (a small vapor app) is here: https://github.com/finestructure/debug-2227

**Comment by finestructure** [comment]
> I may have oversimplified the example and be triggering something that only looks like our issue but it certainly looks like what we're seeing in the full app (which is also using a transaction).

I'm pretty sure now this is our core issue. Luckily it was easy to rearrange the code (in this instance) to pull the slow sync workload out of the transaction and do it up front before the db access. Once I do that, the deadlock errors also disappear in my local reproducer with our actual analysis code.

**Comment by finestructure** [comment]
Looks like this issue is actually two separate issues in a trench coat. After working around the `connection deadlock` error, it (re-)uncovered an underlying crash that is likely the cause for the hang. I've added some more details in our tracking issue and will try to get a reproducer: https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2227#issuecomment-1541400276

**Comment by finestructure** [comment]
Update: the `connection deadlock` errors are also still happening. We've just had it on prod in conjunction with a hang. The earlier report was on `dev` where it hung _without_ any `connection deadlock` errors. Could be that the fix I deployed yesterday just made these error messages less likely and they're unrelated to the hangs.

More details here: https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2227#issuecomment-1541576199

**Comment by finestructure** [comment]
One thing I don't think I mentioned yet: When the hangs happen with the `connection deadlock` error, the process is consuming CPU:

<img width="1072" alt="CleanShot 2023-05-10 at 10 35 19@2x" src="https://github.com/vapor/console-kit/assets/65520/1c5a8a9b-407e-491c-99de-7c8c649fbc5e">

There's no apparent activity though, no logs, no forward process.

When the hang happened earlier on `dev` _without_ the `connection deadlock` message, it hung but the process' CPU was 0%.

**Comment by finestructure** [comment]
Yes, and I managed to work around this particular part of the problem (the `connection deadlock`), I believe.

The real problem for us is that our batch job (the Vapor command) hangs and needs manual restarting. I have my doubts this is actually related to the `connection deadlock` now.

Also note the change in behaviour if the slow job is within a db transaction or not (https://github.com/vapor/console-kit/issues/179#issuecomment-1539516353). Feels like there's something amiss there 🤔

**Comment by finestructure** [comment]
The weird thing is that this timeout seems to be dependent on whether I'm using a db transaction or not (see the comment and the test project above: https://github.com/vapor/console-kit/issues/179#issuecomment-1539516353).

I'll bump the parameter up for now to see if that helps with the frequency of the hangs. Thank you!

**Comment by gwynne** [comment]
Not sure why this is in console-kit; it belongs on async-kit, transferring there.
